### PR TITLE
implemented alternative Try::filter method that accepts supplier for …

### LINKED ISSUE #1273: Try::filter and custom throwable (supplier)
What do think about extending Try with an additional filter method that expects a throwable or throwable supplier? I have a use case where I would like to concretize a problem by a custom Throwable that is more meaningful than a NoSuchElementException if a specific filter is applied. This makes further processing and analyzing of the Try instance easier and more flexible.

If also think that this makes sense I could do a PR.

### PR DESCRIPTION
**Pull request by manu-m**
…providing custom throwable if predicate evaluates to false. fixes #1273

### CODE REVIEW

File: javaslang/src/main/java/javaslang/control/Try.java [hunk 1]
```diff
@@ -238,6 +238,19 @@
...
     /**
+     * Shortcut for {@code filterTry(predicate::test, throwableSupplier)}, see {@link #filterTry(CheckedPredicate)}}.
+     *
+     * @param predicate A predicate
+     * @return a {@code Try} instance
+     * @throws NullPointerException if {@code predicate} or {@code throwableSupplier} is null
+     */
+    default Try<T> filter(Predicate<? super T> predicate, Supplier<? extends Throwable> throwableSupplier) {
+        Objects.requireNonNull(predicate, "predicate is null");
+        Objects.requireNonNull(predicate, "throwableSupplier is null");
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/control/Try.java]: should rather be
```Java
Objects.requireNonNull(throwableSupplier, "throwableSupplier is null");
```

**Comment by paplorinc** [comment]
Nice! :)

If I may suggest, If we've opened up `Try` already, please consider creating some tests for the uncovered methods also https://codecov.io/github/javaslang/javaslang/javaslang/src/main/java/javaslang/control/Try.java?ref=196914af043e5ba2b9860c0aae05ad8a04be2d2a :)

File: javaslang/src/main/java/javaslang/control/Try.java (hunk 1 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/control/Try.java]: good catch

Comment (manu-m) [inline javaslang/src/main/java/javaslang/control/Try.java]: Oh, a little oversight. Thank you for the hint

**Comment by manu-m** [comment]
I will take look at the unit tests in more detail within the next days. I think can do some contribution, although I don't like javaslang's style of unit test at all. I will do a new PR for these contributions.

**Comment by danieldietrich** [comment]
Great, many thanks!

> although I don't like javaslang's style of unit test

Let's improve it where possible...

**Pull request merged**